lib/azurex/blob/block.ex

```diff
@@ -0,0 +1,109 @@
+defmodule Azurex.Blob.Block do
+  @moduledoc """
+  Implementation of Azure Blob Storage.
+
+  You can:
+
+  - [upload a block as part of a blob](https://learn.microsoft.com/en-us/rest/api/storageservices/put-block)
+  - [commit a list of blocks as part of a blob](https://learn.microsoft.com/en-us/rest/api/storageservices/put-block-list)
+  """
+
+  alias Azurex.Authorization.SharedKey
+  alias Azurex.Blob
+  alias Azurex.Blob.Config
+
+  @doc """
+  Creates a block to be committed to a blob.
+
+  On success, returns an :ok tuple with the base64 encoded block_id.
+  """
+  @spec put_block(String.t(), bitstring(), String.t(), list()) ::
+          {:ok, String.t()} | {:error, term()}
+  def put_block(container, chunk, name, params) do
+    block_id = build_block_id()
+    content_type = "application/octet-stream"
+    params = [{:comp, "block"}, {:blockid, block_id} | params]
+
+    %HTTPoison.Request{
+      method: :put,
+      url: Blob.get_url(container, name),
+      params: params,
+      body: chunk,
+      headers: [
+        {"content-type", content_type},
+        {"content-length", byte_size(chunk)}
+      ]
+    }
+    |> SharedKey.sign(
+      storage_account_name: Config.storage_account_name(),
+      storage_account_key: Config.storage_account_key(),
+      content_type: content_type
+    )
+    |> HTTPoison.request()
+    |> case do
+      {:ok, %HTTPoison.Response{status_code: 201}} -> {:ok, block_id}
+      {:ok, err} -> {:error, err}
+      {:error, err} -> {:error, err}
+    end
+  end
+
+  @doc """
+  Commits the given list of block_ids to a blob.
+
+  Block IDs should be base64 encoded, as returned by put_block/2.
+  """
+  @spec put_block_list(list(), String.t(), String.t(), String.t() | nil, list()) ::
+          :ok | {:error, term()}
+  def put_block_list(block_ids, container, name, blob_content_type, params) do
+    params = [{:comp, "blocklist"} | params]
+    content_type = "text/plain; charset=UTF-8"
+    blob_content_type = blob_content_type || "application/octet-stream"
+
+    blocks =
+      block_ids
+      |> Enum.reverse()
+      |> Enum.map(fn block_id -> "<Uncommitted>#{block_id}</Uncommitted>" end)
+      |> Enum.join()
+
+    body = """
+    <?xml version="1.0" encoding="utf-8"?>
+    <BlockList>
+    #{blocks}
+    </BlockList>
+    """
+
+    %HTTPoison.Request{
+      method: :put,
+      url: Blob.get_url(container, name),
+      params: params,
+      body: body,
+      headers: [
+        {"content-type", content_type},
+        {"x-ms-blob-content-type", blob_content_type}
+      ]
+    }
+    |> SharedKey.sign(
+      storage_account_name: Config.storage_account_name(),
+      storage_account_key: Config.storage_account_key(),
+      content_type: content_type
+    )
+    |> HTTPoison.request()
+    |> case do
+      {:ok, %HTTPoison.Response{status_code: 201}} -> :ok
+      {:ok, err} -> {:error, err}
+      {:error, err} -> {:error, err}
+    end
+  end
+
+  defp build_block_id do
```

Comment:
I think this function could be replaced by
```elixir
:rand.bytes(64) |> Base.encode64()
```
So I think, keep the function but replace the implementation with the above

Reply:
I like this idea, but it looks like `:rand.bytes/1` is not available until OTP 24.0, which is only supported by Elixir versions 1.11.4 and up. There is also `:crypto.strong_rand_bytes/1`, but both of these functions create random binaries, which may not be valid strings.

I did notice that I'm padding the generated ID with too many zeroes so it's ending up twice as long as I intended. I'll tighten that up a bit!